torch/autograd/gradcheck.py

```diff
@@ -569,31 +575,24 @@ def fail_test(msg):
                 return False
             numerical_imag_grad_out = get_numerical_jacobian(func, tupled_inputs, i, eps=eps, grad_out=1j)
 
-        def checkIfNumericalAnalyticAreClose(a, n, j, error_str=''):
-            if not torch.allclose(a, n, rtol, atol):
-                return fail_test(error_str + 'Jacobian mismatch for output %d with respect to input %d,\n'
-                                 'numerical:%s\nanalytical:%s\n' % (i, j, n, a))
-
         inp_tensors = iter_tensors(tupled_inputs, True)
 
         for j, (a, n, inp) in enumerate(zip(analytical, numerical, inp_tensors)):
             if a.numel() != 0 or n.numel() != 0:
-                if o.is_complex():
-                    # C -> C, R -> C
-                    a_imag_grad_out = analytical_imag_grad_out[j]
-                    n_imag_grad_out = numerical_imag_grad_out[j]
-                    checkIfNumericalAnalyticAreClose(a_imag_grad_out, n_imag_grad_out, j,
-                                                     "Gradients failed to compare equal for grad output = 1j. ")
-                if inp.is_complex():
-                    # C -> R, C -> C
-                    checkIfNumericalAnalyticAreClose(a, n, j,
-                                                     "Gradients failed to compare equal for grad output = 1. ")
-                else:
-                    # R -> R, R -> C
-                    checkIfNumericalAnalyticAreClose(a, n, j)
+                if o.is_complex():    # C -> C, R -> C
+                    if not torch.allclose(analytical_imag_grad_out[j], numerical_imag_grad_out[j], rtol, atol):
+                        return fail_test(get_notallclose_msg(analytical_imag_grad_out[j], numerical_imag_grad_out[j], i, j,
+                                                             "Gradients failed to compare equal for grad output = 1j. "))
+                if inp.is_complex():  # C -> R, C -> C
+                    if not torch.allclose(a, n, atol, rtol):
+                        return fail_test(get_notallclose_msg(a, n, i, j,
+                                                             "Gradients failed to compare equal for grad output = 1. "))
+                else:                 # R -> R, R -> C
+                    if not torch.allclose(a, n, atol, rtol):
+                        return fail_test(get_notallclose_msg(a, n, i, j))
 
         if check_batched_grad:
-            test_batched_grad(fail_test, tupled_inputs, o, j)
+            return test_batched_grad(fail_test, tupled_inputs, o, i)
```

Comment:
gradcheck should return False if test fails. 
Also j -> i, because we need the output idx. This helps fix the error message.